MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c

```diff
@@ -48,6 +97,210 @@ StatusToHexString (
+  //
+  if ((Attributes & VARIABLE_ATTRIBUTE_NV_BS_RT_AT) != VARIABLE_ATTRIBUTE_NV_BS_RT_AT) {
+    Status = EFI_UNSUPPORTED;
+    //
+    // Likely this will continue to fail on reboot
+    //
+    return FALSE;
+  }
+
+  Status = mBS->AllocatePool (
+                  EfiBootServicesData,
+                  DbSize,
+                  &Db
+                  );
+
+  if ((Db == NULL) || EFI_ERROR (Status)) {
+    //
+    // Likely this will continue to fail on reboot
+    //
+    goto Exit;
+  }
+
+  //
+  // Grab the DB
+  //
+  Status = mRT->GetVariable (
+                  L"db",
+                  &gEfiImageSecurityDatabaseGuid,
+                  &Attributes,
+                  &DbSize,
+                  Db
+                  );
+  if (EFI_ERROR (Status)) {
+    //
+    // Likely this will continue to fail on reboot
+    // GetVariable could return NOT_FOUND if the DB variable does not exist
+    // In that case, we should fail out gracefully because there is nothing we can do
+    //
+    goto Exit;
+  }
+
+  // Confirm the DB is at least as large as the EFI_SIGNATURE_LIST header
+  if (DbSize < sizeof (EFI_SIGNATURE_LIST)) {
+    Status = EFI_BUFFER_TOO_SMALL;
+    goto Exit;
+  }
+
+  //
+  // The DB is a list of EFI_SIGNATURE_LISTs but we only care about certificates
+  // they are not garunteed to be in any particular order
+  //
+  CertList = (EFI_SIGNATURE_LIST *)Db;
+
+  //
+  // Iterate through the DB entries
+  // The DB is a list of EFI_SIGNATURE_LISTs
+  //
+  while ((DbSize > 0) && (DbSize >= CertList->SignatureListSize)) {
+    //
+    // If the signature type is not X509, skip it
+    //
+    if (CompareGuid (&CertList->SignatureType, &gEfiCertX509Guid)) {
+      //
+      // Otherwise, we have a certificate list and now we need to iterate through the certificates
+      //
+      CertData  = (EFI_SIGNATURE_DATA *)((UINT8 *)CertList + sizeof (EFI_SIGNATURE_LIST) + CertList->SignatureHeaderSize);
+      CertCount = (CertList->SignatureListSize - sizeof (EFI_SIGNATURE_LIST) - CertList->SignatureHeaderSize) / CertList->SignatureSize;
+
+      //
+      // Iterate over each certificate in the list
+      //
+      for (Index = 0; Index < CertCount; Index++) {
+        //
+        // grab the certificate and its size
+        //
+        Cert     = CertData->SignatureData;
+        CertSize = CertList->SignatureSize - sizeof (EFI_GUID);
+
+        //
+        // First lets ensure that the size of the certificate is at least as large as the certificate we are looking for
+        // Each offset in the certficate entry plus the size of the string must be less than the size of the certificate
+        //
+        if (((Certificate->OrganizationOffset + sizeof (Certificate->Organization)) <= CertSize) &&
+            ((Certificate->CommonNameOffset + sizeof (Certificate->CommonName)) <= CertSize) &&
+            ((Certificate->SignatureOffset + sizeof (Certificate->Signature)) <= CertSize))

```

Comment:
In `sizeof (Certificate->Organization)` are you trying to get the size of the struct field (i.e. the pointer) or the size of the string itself?

At a quick glance, it seems that you're trying to get the size of the string here. In that case, `AsciiStrSize()` would be needed to get the size including the NULL character.

This is also where making `Organization` being a `CHAR8` makes it compatible with string functions.

Reply:
This is getting the size of the string itself. Although in this example I have OrganizationLength and CommonNameLength and don't need to perform the size_of again.

Since the strings lengths are really just a char array which the size of is known at compile time I figured I could avoid pulling in an additional library to get AsciiStrSize(..) and just use str_len

Further, ASN.1 strings (that we're looking for) aren't null terminated and thus we just need the length of the array

Reply:
`OrganizationLength` and `CommonNameLength` are fine to use because they computed the string length using `sizeof` on a byte-sized string literal.

`sizeof (Certificate->Organization)` is the size of the `Organization` member which is `UINT8*` (8 bytes on 64-bit machine not the underlying string length). `AsciiStrLen()` could be used to not include the null character. But, as you said you can just use the ready-to-go values.